Add get_* and remove_* for extensions

hjmallon opened the pull request:
* unicode: get_keyword, get_attribute, remove_keyword, remove_attribute
* transform: get_tfield, remove_tfield
* private: get_tag, remove_tag

I think I should probably add some documentation to these functions before merging so setting to WIP for now. I would like to hear whether this looks like the right way to go about doing this.

My questions:
* Normal Rust style seems to be remove-and-return from containers. I haven't done this here as we return `Vec<&strs>` from TinyStrs currently, we would have to copy the data out to return it as `Vec<str>` or `Vec<TinyStr>`.
* Is it sensible to return `Result<bool, LocaleError>` from `get_attribute` and similar fns or would it be better to return `Result<Option<()> LocaleError>`?
* I haven't done any tests on the private extensions yet. It looks like it currently is sorted but not deduped. I'm not sure of the semantics of private extensions. Is `-x-foo-foo` a canonical private use extension? Is `-x-zzz-aaa` or should it be `-x-aaa-zzz`? This is probably a question for a different time.
* Should unicode attributes be stored in a BTreeSet to avoid current manual sorting?

zbraniecki commented:
Great patch! Thank you!

I haven't get to review it fully yet, but skimming through it looks like the right direction!
Should we also have some `get_attributes/keywords/tfields` returning an iterator or sth like that?

hjmallon commented:
I have removed the WIP from this as I have added the docs that I wanted to add and I'm happy with it now.

zbraniecki commented on unic-locale-impl/src/extensions/private.rs:
````diff
@@ -16,16 +16,91 @@ fn parse_value(t: &str) -> Result<TinyStr8, ParserError> {
+    /// ```
+    pub fn get_tag(&self, tag: &str) -> Result<bool, LocaleError> {
+        Ok(self.0.contains(&parse_value(tag)?))
+    }
````
Rename to `has_tag`.

zbraniecki replied:
And switch to accepting `AsRef<[u8]>`

zbraniecki commented on unic-locale-impl/src/extensions/private.rs:
````diff
@@ -16,16 +16,91 @@ fn parse_value(t: &str) -> Result<TinyStr8, ParserError> {
+    ///
+    /// assert_eq!(lo.to_string(), "en-US-x-foo");
+    /// ```
     pub fn add_tag(&mut self, tag: &str) -> Result<(), LocaleError> {
````
`AsRef<[u8]>`

zbraniecki commented on unic-locale-impl/src/extensions/private.rs:
````diff
@@ -16,16 +16,91 @@ fn parse_value(t: &str) -> Result<TinyStr8, ParserError> {
+    ///
+    /// assert_eq!(lo.to_string(), "en-US");
+    /// ```
+    pub fn remove_tag(&mut self, tag: &str) -> Result<bool, LocaleError> {
````
`AsRef<[u8]>`

zbraniecki commented on unic-locale-impl/src/extensions/private.rs:
```diff
@@ -16,16 +16,91 @@ fn parse_value(t: &str) -> Result<TinyStr8, ParserError> {
+            Err(_) => Ok(false)
+        }
+    }
+
```
Can you add `clear_tags`?

zbraniecki replied:
And `get_tags` which returns `impl ExactSizeIterator<Item = &str>` (much like `get_variants` does)

zbraniecki replied:
Hmm, maybe name it `get_tag_keys` since I think we should return just keys, not values.

hjmallon replied:
I'll leave `get_tags` for a separate commit. Other thing is resolved.

zbraniecki commented on unic-locale-impl/src/extensions/transform.rs:
````diff
@@ -50,15 +50,94 @@ fn is_language_subtag(t: &str) -> bool {
+    ///
+    /// assert_eq!(lo.to_string(), "en-US-t-es-AR");
+    /// ```
     pub fn set_tlang(&mut self, tlang: LanguageIdentifier) -> Result<(), LocaleError> {
````
Add `get_tlang` and `clear_tlang`.

zbraniecki commented on unic-locale-impl/src/extensions/transform.rs:
````diff
@@ -50,15 +50,94 @@ fn is_language_subtag(t: &str) -> bool {
+    ///                .expect("Getting tfield failed."),
+    ///            None);
+    /// ```
+    pub fn get_tfield(&self, tkey: &str) -> Result<Option<Vec<&str>>, LocaleError> {
````
Hmm, how hard would it be to return `Result<Option<impl ExactSizeIterator>>`? If very hard, I'm ok with going for a `Vec` now.

zbraniecki replied:
But def `AsRef<[u8]>` on input.

zbraniecki commented on unic-locale-impl/src/extensions/transform.rs:
````diff
@@ -50,15 +50,94 @@ fn is_language_subtag(t: &str) -> bool {
+    ///
+    /// assert_eq!(lo.to_string(), "en-US-t-k0-colemak");
+    /// ```
     pub fn set_tfield(&mut self, tkey: &str, tvalue: Vec<&str>) -> Result<(), LocaleError> {
````
`AsRef<[u8]>`

zbraniecki commented on unic-locale-impl/src/extensions/transform.rs:
````diff
@@ -72,6 +151,29 @@ impl TransformExtensionList {
+    ///
+    /// assert_eq!(lo.to_string(), "en-US");
+    /// ```
+    pub fn remove_tfield(&mut self, tkey: &str) -> Result<bool, LocaleError> {
````
`AsRef<[u8]>`

zbraniecki commented on unic-locale-impl/src/extensions/transform.rs:
```diff
@@ -72,6 +151,29 @@ impl TransformExtensionList {
+    pub fn remove_tfield(&mut self, tkey: &str) -> Result<bool, LocaleError> {
+        Ok(self.tfields.remove(&parse_tkey(tkey)?).is_some())
+    }
+
```
Add `clear_tfields` and `get_tfield_keys` which returns `impl ExactSizeIterator`.

hjmallon replied:
Leaving `get_tfield_keys` for separate commit.

zbraniecki commented on unic-locale-impl/src/extensions/unicode.rs:
````diff
@@ -63,10 +63,74 @@ fn is_type(t: &str) -> bool {
+    ///                .expect("Getting keyword failed."),
+    ///            None);
+    /// ```
+    pub fn get_keyword(&self, key: &str) -> Result<Option<Vec<&str>>, LocaleError> {
````
Same question about `Result<Option<impl ExactSizeIterator>>`, and `AsRef<[u8]>`.

zbraniecki commented on unic-locale-impl/src/extensions/unicode.rs:
````diff
@@ -63,10 +63,74 @@ fn is_type(t: &str) -> bool {
+    ///
+    /// assert_eq!(lo.to_string(), "en-US-u-ca-chinese");
+    /// ```
     pub fn set_keyword(&mut self, key: &str, value: Vec<&str>) -> Result<(), LocaleError> {
````
`AsRef<[u8]>` pls

zbraniecki commented on unic-locale-impl/src/extensions/unicode.rs:
````diff
@@ -81,14 +145,100 @@ impl UnicodeExtensionList {
+    ///
+    /// assert_eq!(lo.to_string(), "en-US");
+    /// ```
+    pub fn remove_keyword(&mut self, key: &str) -> Result<bool, LocaleError> {
````
`AsRef<[u8]>`

zbraniecki commented on unic-locale-impl/src/extensions/unicode.rs:
````diff
@@ -81,14 +145,100 @@ impl UnicodeExtensionList {
+    ///                .expect("Getting attribute failed."),
+    ///            true);
+    /// ```
+    pub fn get_attribute(&self, attribute: &str) -> Result<bool, LocaleError> {
````
`AsRef<[u8]>`

zbraniecki replied:
And `has_attribute`.

zbraniecki commented on unic-locale-impl/src/extensions/unicode.rs:
````diff
@@ -81,14 +145,100 @@ impl UnicodeExtensionList {
+    ///
+    /// assert_eq!(lo.to_string(), "en-US-u-foo");
+    /// ```
+    pub fn set_attribute(&mut self, attribute: &str) -> Result<(), LocaleError> {
````
`AsRef<[u8]>`

zbraniecki commented on unic-locale-impl/src/extensions/unicode.rs:
````diff
@@ -81,14 +145,100 @@ impl UnicodeExtensionList {
+    ///
+    /// assert_eq!(lo.to_string(), "en-US");
+    /// ```
+    pub fn remove_attribute(&mut self, attribute: &str) -> Result<bool, LocaleError> {
````
`AsRef<[u8]>`

zbraniecki commented on unic-locale-impl/src/extensions/unicode.rs:
```diff
@@ -81,14 +145,100 @@ impl UnicodeExtensionList {
+            Err(_) => Ok(false)
+        }
+    }
+
```
`get_attribute_keys` and `clear_attributes`/`clear_keywords`.

hjmallon replied:
Leaving `get_attribute_keys` for separate commit too.

zbraniecki commented:
Oh, and let me know if you have no time soon, I want to release 0.7 soon and am happy to take over from you if you're busy! :)

hjmallon commented:
Hi. I have:

* Added clear fns
* Added tlang fns
* Rebased to master and used AsRef<[u8]>s
* Changed the return type of `get_keyword` and `get_tfield` to use ExactSizeIterator. However I am completely unclear on how this works and basically copied the `variants` version blindly. !!

I have not:
* Implemented `get_attributes`, `get_keyword_keys`, `get_tfield_tkeys`, `get_tags` fns which return iterators. I ran out of time and this can be a separate commit.

hjmallon commented:
I have now added the `get_tags`, `get_attributes`, `get_keyword_keys` and `get_tfield_tkeys` functions returning iterators.
